vulnerability/events_creator.go

```diff
@@ -94,13 +90,20 @@ type Resource struct {
 }
 
 const (
-	vulScannerVersion     = "v0.35.0"
-	vulScannerVendor      = "Trivy"
-	vulScoreSystemClass   = "CVSS"
-	vulScoreSource        = "nvd"
-	vulScoreSourceVersion = "3.0"
-	vulEcsCategory        = "vulnerability"
-	vulIndex              = "logs-cloud_security_posture.vulnerabilities-default"
+	vulScannerVersion   = "v0.35.0"
+	vulScannerVendor    = "Trivy"
+	vulScoreSystemClass = "CVSS"
+	nvdSource           = "nvd"
+	ghsaSource          = "ghsa"
+	redhatSource        = "redhat"
+	vectorHeader        = "CVSS:"
+	vulEcsCategory      = "vulnerability"
+	vulIndex            = "logs-cloud_security_posture.vulnerabilities-default"
+)
+
+var (
+	// sources order is crucial as it defines the priority
+	sources = []db_types.SourceID{nvdSource, redhatSource, ghsaSource}
```

Comment:
I think that we should maybe try and pull the sources from Trivy in order to have a more dynamic code and one that is prone to the addition and removal of sources from Trivy's side without us needing to change our code.
A quick look at their source base, I saw that they list them and it looks like they list them:
```golang
package vulnerability

import "github.com/aquasecurity/trivy-db/pkg/types"

const (
	// Data source
	NVD                   types.SourceID = "nvd"
	RedHat                types.SourceID = "redhat"
	RedHatOVAL            types.SourceID = "redhat-oval"
	Debian                types.SourceID = "debian"
	Ubuntu                types.SourceID = "ubuntu"
	CentOS                types.SourceID = "centos"
	Rocky                 types.SourceID = "rocky"
	Fedora                types.SourceID = "fedora"
	Amazon                types.SourceID = "amazon"
	OracleOVAL            types.SourceID = "oracle-oval"
	SuseCVRF              types.SourceID = "suse-cvrf"
	Alpine                types.SourceID = "alpine"
	ArchLinux             types.SourceID = "arch-linux"
	Alma                  types.SourceID = "alma"
	CBLMariner            types.SourceID = "cbl-mariner"
	Photon                types.SourceID = "photon"
	RubySec               types.SourceID = "ruby-advisory-db"
	PhpSecurityAdvisories types.SourceID = "php-security-advisories"
	NodejsSecurityWg      types.SourceID = "nodejs-security-wg"
	GHSA                  types.SourceID = "ghsa"
	GLAD                  types.SourceID = "glad"
	GoVulnDB              types.SourceID = "go-vulndb"
	OSV                   types.SourceID = "osv"
	Wolfi                 types.SourceID = "wolfi"
	Chainguard            types.SourceID = "chainguard"

	// Ecosystem
	Npm      types.Ecosystem = "npm"
	Composer types.Ecosystem = "composer"
	Pip      types.Ecosystem = "pip"
	RubyGems types.Ecosystem = "rubygems"
	Cargo    types.Ecosystem = "cargo"
	NuGet    types.Ecosystem = "nuget"
	Maven    types.Ecosystem = "maven"
	Go       types.Ecosystem = "go"
	Rust     types.Ecosystem = "rust"
	Conan    types.Ecosystem = "conan"
	Erlang   types.Ecosystem = "erlang"
	Pub      types.Ecosystem = "pub"
)
```

Reply:
Notice this source code from Trivy as well:
```golang
// FillInfo fills extra info in vulnerability objects
func (c Client) FillInfo(vulns []types.DetectedVulnerability) {
	for i := range vulns {
		vulnID := vulns[i].VulnerabilityID
		vuln, err := c.dbc.GetVulnerability(vulnID)
		if err != nil {
			log.Logger.Warnf("Error while getting vulnerability details: %s", err)
			continue
		}

		// Detect the data source
		var source dbTypes.SourceID
		if vulns[i].DataSource != nil {
			source = vulns[i].DataSource.ID
		}

		// Select the severity according to the detected source.
		severity, severitySource := c.getVendorSeverity(vulnID, &vuln, source)

		// The vendor might provide package-specific severity like Debian.
		// For example, CVE-2015-2328 in Debian has "unimportant" for mongodb and "low" for pcre3.
		// In that case, we keep the severity as is.
		if vulns[i].SeveritySource != "" {
			severity = vulns[i].Severity
			severitySource = vulns[i].SeveritySource
		}

		// Add the vulnerability detail
		vulns[i].Vulnerability = vuln

		vulns[i].Severity = severity
		vulns[i].SeveritySource = severitySource
		vulns[i].PrimaryURL = c.getPrimaryURL(vulnID, vuln.References, source)
	}
}

func (c Client) getVendorSeverity(vulnID string, vuln *dbTypes.Vulnerability, source dbTypes.SourceID) (string, dbTypes.SourceID) {
	if vs, ok := vuln.VendorSeverity[source]; ok {
		return vs.String(), source
	}

	// use severity from GitHub for all GHSA-xxx vulnerabilities
	// govuln doesn't have severity. Use severity from GitHub(for CVE-xxx vulns)
	if strings.HasPrefix(vulnID, "GHSA-") || source == vulnerability.GoVulnDB {
		if vs, ok := vuln.VendorSeverity[vulnerability.GHSA]; ok {
			return vs.String(), vulnerability.GHSA
		}
	}

	// Try NVD as a fallback if it exists
	if vs, ok := vuln.VendorSeverity[vulnerability.NVD]; ok {
		return vs.String(), vulnerability.NVD
	}

	if vuln.Severity == "" {
		return dbTypes.SeverityUnknown.String(), ""
	}

	return vuln.Severity, ""
}
```

Reply:
It appears that relying on the logic implemented by Trivy and using the data source as the primary option would be the better approach, with the NVD fields being used as a fallback.